polygeist/lib/Dialect/Polygeist/Analysis/SYCLIDAndRangeAnalysis.cpp

```diff
@@ -0,0 +1,241 @@
+  return *constantValues;
+}
+
+const IDRangeInformation
+IDRangeInformation::join(const IDRangeInformation &other) const {
+  if (isConstant() && other.isConstant()) {
+    if (getConstantValues() == other.getConstantValues())
+      return *this;
+  }
+  if (hasFixedDimensions() && other.hasFixedDimensions()) {
+    if (getNumDimensions() == other.getNumDimensions())
+      return IDRangeInformation{getNumDimensions()};
+  }
+  return IDRangeInformation{};
+}
+
+//===----------------------------------------------------------------------===//
+// Helpers
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+static std::optional<int> getConstantUInt(Value v) {
+  Operation *op = v.getDefiningOp();
+  if (!op)
+    return std::nullopt;
+
+  if (!op->hasTrait<OpTrait::ConstantLike>())
+    return std::nullopt;
+
+  llvm::SmallVector<OpFoldResult> folded;
+  if (failed(op->fold({}, folded)) || folded.size() != 1)
+    return std::nullopt;
+
+  if (!folded.front().is<Attribute>() ||
+      !isa<IntegerAttr>(folded.front().get<Attribute>()))
+    return std::nullopt;
+
+  return cast<IntegerAttr>(folded.front().get<Attribute>()).getInt();
+}
+
+template <typename Type> bool isConstructor(const Definition &def) {
+  if (!def.isOperation())
+    return false;
+
+  auto constructor = dyn_cast<sycl::SYCLHostConstructorOp>(def.getOperation());
+  if (!constructor)
+    return false;
+
+  return isa<Type>(constructor.getType().getValue());
+}
+
+template <typename IDRange>
+IDRangeInformation getInformation(const Definition &def) {
+  assert(def.isOperation() && "Expecting operation");
+
+  auto constructor = cast<sycl::SYCLHostConstructorOp>(def.getOperation());
+
+  auto type = cast<IDRange>(constructor.getType().getValue());
+
+  SmallVector<std::optional<int>> constValues;
+  llvm::transform(constructor.getArgs(), std::back_inserter(constValues),
+                  getConstantUInt);
+
+  if (llvm::all_of(constValues, [](auto &opt) { return opt.has_value(); })) {
+    SmallVector<size_t, 3> constInt;
+    llvm::transform(constValues, std::back_inserter(constInt),
+                    [](auto &opt) { return *opt; });
+    return IDRangeInformation(constInt);
+  }
+
+  return IDRangeInformation(type.getDimension());
+}
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// SYCLIDAndRangeAnalysis
+//===----------------------------------------------------------------------===//
+
+SYCLIDAndRangeAnalysis::SYCLIDAndRangeAnalysis(Operation *op,
+                                               AnalysisManager &mgr)
+    : operation(op), am(mgr) {
+  initialize();
+}
+
+void SYCLIDAndRangeAnalysis::initialize() {
+
+  // Initialize the dataflow solver
+  AliasAnalysis &aliasAnalysis = am.getAnalysis<mlir::AliasAnalysis>();
+  aliasAnalysis.addAnalysisImplementation(
+      sycl::AliasAnalysis(/* relaxedAliasing */ false));
```

Comment:
This would be difficult with the current interface. The current interface is the interface required for analysis that are handled by the analysis manager. 

If we want the relaxed aliasing to be controlled by the users of the analysis, we would probably need an interface similar to [`ReachingDefinitionAnalysis`](https://github.com/sommerlukas/llvm/blob/id-range-analysis/polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h#L135). 

If you compare the [test pass for reaching definition](https://github.com/sommerlukas/llvm/blob/id-range-analysis/polygeist/test/lib/Dialect/Polygeist/TestReachingDefinitionsAnalysis.cpp#L37) with the [one for this analysis](https://github.com/sommerlukas/llvm/blob/id-range-analysis/polygeist/test/lib/Dialect/Polygeist/TestIDAndRangeAnalysis.cpp#L25) you can see the difference in usage. 

If you prefer to have control over the relaxed aliasing behavior, we could switch to that.

Reply:
I think that would be nice, pass `bool relaxedAliasing` to `initialize`. Similar to what is done in `MemoryAccessAnalysis` ?